python/pyspark/ml/torch/deepspeed/deepspeed_distributer.py

```diff
@@ -0,0 +1,167 @@
+            )
+        return gpu_mapped_to_node
+
+    def _setup_hostfile_info(self):
+        worker_hosts = [
+            executor.host()
+            for executor in self.spark.sparkContext._jsc.sc().statusTracker().getExecutorInfos()
+        ]  # we should check if this returns the driver or not
+        worker_count = len(worker_hosts)  # find out if this number includes the driver or not
+        rdd = spark.sparkContext.parallelize(range(worker_count), numSlices=worker_count)
+
+        # what do I do if the use_gpu flag is false?
+        slots_on_workers = {}
+        if self.use_gpu:
+            for worker_host in worker_hosts:
+                slots_on_workers[worker_host] = self._get_gpus_on_node(worker_host)
+        else:
+            raise RuntimeError("Deepspeed doesn't work with non-GPU clusters at this time")
+
+        assigned_slots = self._assign_procs_to_worker(slots_on_workers)
+        self.logger.info(f"Writing to {DeepspeedDistributor.HOSTFILE}")
+        for worker_host in worker_hosts:
+            line = f"{worker_host} slots={assigned_slots[worker_host]}\n"
+            write_to_location(DeepspeedDistributor.HOSTFILE, line)
+        return worker_hosts
+
+    def setup_env(self):
+        try:
+            subprocess.run("deepspeed --version".split())
+            subprocess.run("ninja --version".split())
+            with open(f"/{DeepspeedDistributor.HOME}/.deepspeed_env", "w") as f:
+                # if this is open; don't add that to path if they're not running on databricks
+                # TODO: figure out what paths to add to this, because if this is OSS we don't want to constantly add a databricks filepath
+                f.write(
+                    "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"
+                )
+        except:
+            raise ImportError("Install deepspeed and ninja on the cluster using PyPi")
+
+    def _create_deepspeed_command(
+        self, input_params: Dict[str, Any], path_to_train_file: str, *args: Any
+    ):
+        local_mode = input_params["local_mode"]
+        num_processes = input_params["num_processes"]
+        deepspeed_config = input_params["deepspeed_config"]
+        if isinstance(deepspeed_config, dict):
+            with tempfile.NamedTemporaryFile(mode="w+", delete=False, suffix=".json") as f:
+                json.dump(deepspeed_config, f)
+                deepspeed_config_path = f.name
+                self.temp_deepspeed_fname = f.name
+        else:
+            deepspeed_config_path = deepspeed_config
+        if local_mode:
+            deepspeed_args = [
+                "--num_gpus",
+                str(num_processes),
+            ]  # no need for num nodes, the host file, or any port stuff (no communiation)
+        else:
+            deepspeed_args = [
+                "--num_gpus",
+                str(input_params["num_processes"]),
```

Comment:
Does user pass `num_processes` as (a) the total number of GPUs among worker nodes, or (b) the max number of GPUs on each node?

`num_processes` for `TorchDistributor` is (a), meanwhile Deepspeed launcher takes `num_gpus` as (b): https://github.com/microsoft/DeepSpeed/blob/master/deepspeed/launcher/runner.py#L105